{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "90da34b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os \n",
    "from functions import *\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be11eda1",
   "metadata": {},
   "source": [
    "# Gene-locustags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9d46d93d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_features = pd.read_csv('../data/genome/GCF_000203855.3_ASM20385v3_feature_table.txt',sep='\\t')\n",
    "ref_features = ( ref_features[ref_features['# feature']=='gene']).reset_index().drop(['index'],axis=1)\n",
    "ref_features = ref_features[['start','end','strand','feature_interval_length','symbol','attributes']]\n",
    "ref_features = ( ref_features.dropna(subset=['attributes']) ).reset_index().drop(['index'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46639b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>symbol</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1368</td>\n",
       "      <td>+</td>\n",
       "      <td>1368</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>lp_0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1546</td>\n",
       "      <td>2685</td>\n",
       "      <td>+</td>\n",
       "      <td>1140</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>lp_0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3210</td>\n",
       "      <td>3443</td>\n",
       "      <td>+</td>\n",
       "      <td>234</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>lp_0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3444</td>\n",
       "      <td>4568</td>\n",
       "      <td>+</td>\n",
       "      <td>1125</td>\n",
       "      <td>recF</td>\n",
       "      <td>lp_0005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4565</td>\n",
       "      <td>6511</td>\n",
       "      <td>+</td>\n",
       "      <td>1947</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>lp_0006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end strand  feature_interval_length symbol locus_tag\n",
       "0      1  1368      +                     1368   dnaA   lp_0001\n",
       "1   1546  2685      +                     1140   dnaN   lp_0002\n",
       "2   3210  3443      +                      234   yaaA   lp_0004\n",
       "3   3444  4568      +                     1125   recF   lp_0005\n",
       "4   4565  6511      +                     1947   gyrB   lp_0006"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "locus_tag=[]\n",
    "for i in range(len(ref_features.index)):\n",
    "    att = list(ref_features['attributes'])[i]\n",
    "    if 'old_locus_tag=' in str( att ):\n",
    "        locus_tag.append( str(att.split('=')[1]).strip()  )\n",
    "    else:\n",
    "        locus_tag.append('-')\n",
    "ref_features['locus_tag']=locus_tag\n",
    "ref_features = ref_features.drop(['attributes'], axis=1)\n",
    "ref_features['symbol'] = ref_features['symbol'].fillna(ref_features['locus_tag'])\n",
    "ref_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fe7ea16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>symbol</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1368</td>\n",
       "      <td>+</td>\n",
       "      <td>1368</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>lp_0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1546</td>\n",
       "      <td>2685</td>\n",
       "      <td>+</td>\n",
       "      <td>1140</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>lp_0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3210</td>\n",
       "      <td>3443</td>\n",
       "      <td>+</td>\n",
       "      <td>234</td>\n",
       "      <td>lp_0004</td>\n",
       "      <td>lp_0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3444</td>\n",
       "      <td>4568</td>\n",
       "      <td>+</td>\n",
       "      <td>1125</td>\n",
       "      <td>recF</td>\n",
       "      <td>lp_0005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4565</td>\n",
       "      <td>6511</td>\n",
       "      <td>+</td>\n",
       "      <td>1947</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>lp_0006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end strand  feature_interval_length   symbol locus_tag\n",
       "0      1  1368      +                     1368     dnaA   lp_0001\n",
       "1   1546  2685      +                     1140     dnaN   lp_0002\n",
       "2   3210  3443      +                      234  lp_0004   lp_0004\n",
       "3   3444  4568      +                     1125     recF   lp_0005\n",
       "4   4565  6511      +                     1947     gyrB   lp_0006"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb_features = pd.read_csv('../data/genome/GCA_000203855.3_ASM20385v3_feature_table.txt',sep='\\t')\n",
    "gb_features = (gb_features[gb_features['# feature']=='gene']).reset_index().drop(['index'],axis=1)\n",
    "gb_features = gb_features[['start','end','strand','feature_interval_length','symbol','locus_tag']]\n",
    "gb_features = ( gb_features.dropna(subset=['locus_tag']) ).reset_index().drop(['index'],axis=1)\n",
    "gb_features['symbol'] = gb_features['symbol'].fillna(gb_features['locus_tag'])\n",
    "gb_features.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "113d7f59",
   "metadata": {},
   "source": [
    "# Lactobacillus plantarum cauh2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2456555",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_clean_fasta( in_path, out_path ):\n",
    "    temp_records = SeqIO.parse( in_path , \"fasta\")\n",
    "    new_records = []\n",
    "    for seq_record in temp_records:\n",
    "        description = str(seq_record.description)\n",
    "        if 'locus_tag=' in description:\n",
    "            ID = str(seq_record.description).split('locus_tag=')[1].split('] [')[0].strip()\n",
    "        elif 'gene=' in description:\n",
    "            ID = str(seq_record.description).split('gene=')[1].split('] [')[0].strip()\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "        seq = str(seq_record.seq)\n",
    "        new_records.append( SeqRecord(Seq(seq), id = ID, name=\"\",description=\"\") )\n",
    "    SeqIO.write(new_records, out_path ,\"fasta\")\n",
    "    print( len(new_records) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4169a7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save_clean_fasta( \"../data/genome/GCA_000203855.3_ASM20385v3_cds_from_genomic.fna\",\n",
    "#                  \"../data/genome/wcfs1_cds.fna\")\n",
    "# save_clean_fasta( \"../data/genome/GCA_001617525.1_ASM161752v1_cds_from_genomic.fna\",\n",
    "#                  \"../data/genome/cauh2_cds.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2100557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cmd = 'blastn -query ../data/genome/cauh2_cds.fna -subject ../data/genome/wcfs1_cds.fna -evalue 1e-6 -outfmt 6'\n",
    "# cmd += ' -max_target_seqs 5 -out ../data/genome/cauh2vswcfs1.txt'\n",
    "# os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "637fdb9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_table = pd.read_csv('../data/genome/cauh2vswcfs1.txt',sep='\\t',header=None)\n",
    "blast_table = blast_table.rename({0:'cauh2_id',1:'wcfs1_id',10:'evalue',11:'bitscore'},axis=1)\n",
    "cauh2_genes = list(np.unique(blast_table['cauh2_id']))\n",
    "print(len(cauh2_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d68cab9",
   "metadata": {},
   "source": [
    "# Combine transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2300b0",
   "metadata": {},
   "source": [
    "## L. plantarum treated with 3OC12 at three time points(Convert count to RPKM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2b73f648",
   "metadata": {},
   "outputs": [],
   "source": [
    "GSE124050_ADTV7 = pd.read_csv('../data/transcriptomics/GSE124050_ADTV7.txt',sep='\\t')\n",
    "GSE124050_AYR0M = pd.read_csv('../data/transcriptomics/GSE124050_AYR0M.txt',sep='\\t')\n",
    "GSE124050 = GSE124050_ADTV7.merge( GSE124050_AYR0M, on=['gene'],how='inner')\n",
    "GSE124050 = GSE124050.rename(columns={'gene':'locus_tag'})\n",
    "sample_list = list(GSE124050.columns)[1:]\n",
    "rename_dict = {x: 'wt_3OC_'+x.split('_')[0].strip() for x in sample_list}\n",
    "GSE124050 = GSE124050.rename(columns = rename_dict )\n",
    "sample_list = list(GSE124050.columns)[1:]\n",
    "totalReads = {sample: sum(GSE124050[sample]) for sample in sample_list }\n",
    "drop_inds = []\n",
    "for i in range(len( GSE124050.index )):\n",
    "    if ( list(GSE124050['locus_tag'])[i] not in list(ref_features['locus_tag']) ) and \\\n",
    "        ( list(GSE124050['locus_tag'])[i] not in list(gb_features['locus_tag']) ):\n",
    "            drop_inds.append(i)\n",
    "GSE124050 = ( GSE124050.drop(drop_inds) ).reset_index().drop(['index'],axis=1)  \n",
    "lengths = []\n",
    "for i in range(len( GSE124050.index )):\n",
    "    gene_id = list(GSE124050['locus_tag'])[i]\n",
    "    if gene_id in list(ref_features['locus_tag']):\n",
    "        temp = ref_features[ref_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )\n",
    "    else:\n",
    "        temp = gb_features[gb_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )       \n",
    "GSE124050['length'] = lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "320c3631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3138, 37)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>wt_3OC_1m1b</th>\n",
       "      <th>wt_3OC_1m4b</th>\n",
       "      <th>wt_3OC_1m7b</th>\n",
       "      <th>wt_3OC_1p1b</th>\n",
       "      <th>wt_3OC_1p4b</th>\n",
       "      <th>wt_3OC_1p7b</th>\n",
       "      <th>wt_3OC_2m1b</th>\n",
       "      <th>wt_3OC_2m4b</th>\n",
       "      <th>wt_3OC_2m7b</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_3OC_2m7</th>\n",
       "      <th>wt_3OC_2p1</th>\n",
       "      <th>wt_3OC_2p4</th>\n",
       "      <th>wt_3OC_2p7</th>\n",
       "      <th>wt_3OC_3m1</th>\n",
       "      <th>wt_3OC_3m4</th>\n",
       "      <th>wt_3OC_3m7</th>\n",
       "      <th>wt_3OC_3p1</th>\n",
       "      <th>wt_3OC_3p4</th>\n",
       "      <th>wt_3OC_3p7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>411.346138</td>\n",
       "      <td>553.351197</td>\n",
       "      <td>249.371632</td>\n",
       "      <td>209.449487</td>\n",
       "      <td>481.504846</td>\n",
       "      <td>311.347276</td>\n",
       "      <td>402.061680</td>\n",
       "      <td>430.104200</td>\n",
       "      <td>333.087531</td>\n",
       "      <td>...</td>\n",
       "      <td>333.273715</td>\n",
       "      <td>195.250348</td>\n",
       "      <td>543.782304</td>\n",
       "      <td>348.974973</td>\n",
       "      <td>346.797211</td>\n",
       "      <td>635.913367</td>\n",
       "      <td>341.119281</td>\n",
       "      <td>196.448803</td>\n",
       "      <td>448.767525</td>\n",
       "      <td>309.455028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>675.903244</td>\n",
       "      <td>729.459234</td>\n",
       "      <td>466.167423</td>\n",
       "      <td>350.164700</td>\n",
       "      <td>651.312610</td>\n",
       "      <td>557.832936</td>\n",
       "      <td>604.242954</td>\n",
       "      <td>594.243364</td>\n",
       "      <td>548.100832</td>\n",
       "      <td>...</td>\n",
       "      <td>502.474216</td>\n",
       "      <td>328.605421</td>\n",
       "      <td>665.000442</td>\n",
       "      <td>592.367716</td>\n",
       "      <td>584.858376</td>\n",
       "      <td>2688.804675</td>\n",
       "      <td>578.271646</td>\n",
       "      <td>335.371467</td>\n",
       "      <td>541.017485</td>\n",
       "      <td>515.892220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>376.813991</td>\n",
       "      <td>357.391046</td>\n",
       "      <td>613.380595</td>\n",
       "      <td>273.134070</td>\n",
       "      <td>540.608405</td>\n",
       "      <td>644.894079</td>\n",
       "      <td>377.668864</td>\n",
       "      <td>367.322540</td>\n",
       "      <td>501.373205</td>\n",
       "      <td>...</td>\n",
       "      <td>576.440654</td>\n",
       "      <td>422.038793</td>\n",
       "      <td>424.975166</td>\n",
       "      <td>558.684419</td>\n",
       "      <td>347.935549</td>\n",
       "      <td>124.605067</td>\n",
       "      <td>418.748057</td>\n",
       "      <td>455.054122</td>\n",
       "      <td>434.364899</td>\n",
       "      <td>615.336113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>265.064973</td>\n",
       "      <td>257.563170</td>\n",
       "      <td>359.618266</td>\n",
       "      <td>269.134277</td>\n",
       "      <td>250.318908</td>\n",
       "      <td>352.331061</td>\n",
       "      <td>288.394153</td>\n",
       "      <td>264.653997</td>\n",
       "      <td>331.214275</td>\n",
       "      <td>...</td>\n",
       "      <td>352.105323</td>\n",
       "      <td>310.763012</td>\n",
       "      <td>234.571574</td>\n",
       "      <td>342.207692</td>\n",
       "      <td>241.756904</td>\n",
       "      <td>359.569214</td>\n",
       "      <td>297.949608</td>\n",
       "      <td>317.757793</td>\n",
       "      <td>243.216544</td>\n",
       "      <td>359.557490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>433.302099</td>\n",
       "      <td>440.621223</td>\n",
       "      <td>381.125583</td>\n",
       "      <td>452.061103</td>\n",
       "      <td>379.905628</td>\n",
       "      <td>350.904239</td>\n",
       "      <td>424.262665</td>\n",
       "      <td>471.276575</td>\n",
       "      <td>399.911122</td>\n",
       "      <td>...</td>\n",
       "      <td>424.914035</td>\n",
       "      <td>467.419220</td>\n",
       "      <td>398.212777</td>\n",
       "      <td>369.068335</td>\n",
       "      <td>409.124564</td>\n",
       "      <td>217.431236</td>\n",
       "      <td>440.238369</td>\n",
       "      <td>482.840170</td>\n",
       "      <td>383.178108</td>\n",
       "      <td>357.881465</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  wt_3OC_1m1b  wt_3OC_1m4b  wt_3OC_1m7b  wt_3OC_1p1b  wt_3OC_1p4b  \\\n",
       "0   lp_0001   411.346138   553.351197   249.371632   209.449487   481.504846   \n",
       "1   lp_0002   675.903244   729.459234   466.167423   350.164700   651.312610   \n",
       "2   lp_0004   376.813991   357.391046   613.380595   273.134070   540.608405   \n",
       "3   lp_0005   265.064973   257.563170   359.618266   269.134277   250.318908   \n",
       "4   lp_0006   433.302099   440.621223   381.125583   452.061103   379.905628   \n",
       "\n",
       "   wt_3OC_1p7b  wt_3OC_2m1b  wt_3OC_2m4b  wt_3OC_2m7b  ...  wt_3OC_2m7  \\\n",
       "0   311.347276   402.061680   430.104200   333.087531  ...  333.273715   \n",
       "1   557.832936   604.242954   594.243364   548.100832  ...  502.474216   \n",
       "2   644.894079   377.668864   367.322540   501.373205  ...  576.440654   \n",
       "3   352.331061   288.394153   264.653997   331.214275  ...  352.105323   \n",
       "4   350.904239   424.262665   471.276575   399.911122  ...  424.914035   \n",
       "\n",
       "   wt_3OC_2p1  wt_3OC_2p4  wt_3OC_2p7  wt_3OC_3m1   wt_3OC_3m4  wt_3OC_3m7  \\\n",
       "0  195.250348  543.782304  348.974973  346.797211   635.913367  341.119281   \n",
       "1  328.605421  665.000442  592.367716  584.858376  2688.804675  578.271646   \n",
       "2  422.038793  424.975166  558.684419  347.935549   124.605067  418.748057   \n",
       "3  310.763012  234.571574  342.207692  241.756904   359.569214  297.949608   \n",
       "4  467.419220  398.212777  369.068335  409.124564   217.431236  440.238369   \n",
       "\n",
       "   wt_3OC_3p1  wt_3OC_3p4  wt_3OC_3p7  \n",
       "0  196.448803  448.767525  309.455028  \n",
       "1  335.371467  541.017485  515.892220  \n",
       "2  455.054122  434.364899  615.336113  \n",
       "3  317.757793  243.216544  359.557490  \n",
       "4  482.840170  383.178108  357.881465  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for sample in sample_list:\n",
    "    temp_rpkm = [ get_rpkm( list(GSE124050[sample])[i],\n",
    "                    list(GSE124050['length'])[i], totalReads[sample]) for i in range(len(GSE124050.index))]\n",
    "    GSE124050[sample] = temp_rpkm\n",
    "GSE124050 = GSE124050[['locus_tag']+sample_list]\n",
    "print(GSE124050.shape)\n",
    "GSE124050.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16e05a3",
   "metadata": {},
   "source": [
    "## L. plantarum strains from nine different habitats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0200fc2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3108, 37)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>WCFS1_MRS</th>\n",
       "      <th>LB16_MRS</th>\n",
       "      <th>TJ22</th>\n",
       "      <th>TJ20</th>\n",
       "      <th>TJ19</th>\n",
       "      <th>WFH19</th>\n",
       "      <th>WFH20</th>\n",
       "      <th>WFH1W</th>\n",
       "      <th>PJ16</th>\n",
       "      <th>...</th>\n",
       "      <th>CB5</th>\n",
       "      <th>FEV11_MRS</th>\n",
       "      <th>FEO3_MRS</th>\n",
       "      <th>DE12_MRS</th>\n",
       "      <th>CB5_MRS</th>\n",
       "      <th>OE1_MRS</th>\n",
       "      <th>WFH19_MRS</th>\n",
       "      <th>TJ22_MRS</th>\n",
       "      <th>PJ16_MRS</th>\n",
       "      <th>BEE1ST_MRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>repA</td>\n",
       "      <td>190.5490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.282865</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>orf2</td>\n",
       "      <td>59.9286</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>246.269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>33.7182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.90913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>orf3</td>\n",
       "      <td>116.1010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>repB</td>\n",
       "      <td>357.2480</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.361062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>orf2</td>\n",
       "      <td>203.2870</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  WCFS1_MRS  LB16_MRS  TJ22  TJ20      TJ19     WFH19  WFH20  \\\n",
       "0      repA   190.5490       0.0   0.0   0.0  0.000000  0.282865    0.0   \n",
       "1      orf2    59.9286       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "2      orf3   116.1010       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "3      repB   357.2480       0.0   0.0   0.0  0.361062  0.000000    0.0   \n",
       "4      orf2   203.2870       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "\n",
       "     WFH1W  PJ16  ...      CB5  FEV11_MRS  FEO3_MRS  DE12_MRS  CB5_MRS  \\\n",
       "0    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "1  246.269   0.0  ...  33.7182        0.0       0.0       0.0  3.90913   \n",
       "2    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "3    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "4    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "\n",
       "   OE1_MRS  WFH19_MRS  TJ22_MRS  PJ16_MRS  BEE1ST_MRS  \n",
       "0      0.0        0.0       0.0       0.0         0.0  \n",
       "1      0.0        0.0       0.0       0.0         0.0  \n",
       "2      0.0        0.0       0.0       0.0         0.0  \n",
       "3      0.0        0.0       0.0       0.0         0.0  \n",
       "4      0.0        0.0       0.0       0.0         0.0  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE115448 = pd.read_csv('../data/transcriptomics/GSE115448_RPKMvalues.txt',sep='\\t')\n",
    "sample_list = list(GSE115448.columns)[2:]\n",
    "IDs = []\n",
    "for i in range(len(GSE115448.index)):\n",
    "    if 'lp_' in list(GSE115448['GeneSymbol'])[i]:\n",
    "        IDs.append( list(GSE115448['GeneSymbol'])[i].split(':')[1].strip() )\n",
    "    elif 'lp_' in list(GSE115448['GeneName'])[i]:\n",
    "        IDs.append( list(GSE115448['GeneName'])[i].strip() )\n",
    "    elif list(GSE115448['GeneName'])[i] in list(ref_features['symbol']):\n",
    "        temp = ref_features[ref_features['symbol'] == list(GSE115448['GeneName'])[i] ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    elif list(GSE115448['GeneName'])[i] in list(gb_features['symbol']):\n",
    "        temp = gb_features[gb_features['symbol'] == list(GSE115448['GeneName'])[i] ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    else:\n",
    "        IDs.append(list(GSE115448['GeneName'])[i])\n",
    "GSE115448['locus_tag'] = IDs\n",
    "GSE115448 = GSE115448.drop(['GeneSymbol','GeneName'], axis = 1)\n",
    "GSE115448 = GSE115448[['locus_tag']+sample_list]\n",
    "print(GSE115448.shape)\n",
    "GSE115448.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c4cd240",
   "metadata": {},
   "source": [
    "## L. plantarum response to acidic conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b9c586c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3174, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>wt_pH6.2_1</th>\n",
       "      <th>wt_pH6.2_2</th>\n",
       "      <th>wt_pH5.5_1</th>\n",
       "      <th>wt_pH5.5_2</th>\n",
       "      <th>wt_pH5.0_1</th>\n",
       "      <th>wt_pH5.0_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>423.10</td>\n",
       "      <td>406.42</td>\n",
       "      <td>468.07</td>\n",
       "      <td>465.00</td>\n",
       "      <td>619.47</td>\n",
       "      <td>640.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>672.64</td>\n",
       "      <td>645.10</td>\n",
       "      <td>681.56</td>\n",
       "      <td>685.90</td>\n",
       "      <td>897.24</td>\n",
       "      <td>916.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>510.91</td>\n",
       "      <td>559.08</td>\n",
       "      <td>522.82</td>\n",
       "      <td>514.83</td>\n",
       "      <td>499.21</td>\n",
       "      <td>448.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>358.58</td>\n",
       "      <td>345.58</td>\n",
       "      <td>294.79</td>\n",
       "      <td>299.09</td>\n",
       "      <td>288.61</td>\n",
       "      <td>290.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>437.95</td>\n",
       "      <td>431.14</td>\n",
       "      <td>357.30</td>\n",
       "      <td>361.36</td>\n",
       "      <td>354.50</td>\n",
       "      <td>360.22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  wt_pH6.2_1  wt_pH6.2_2  wt_pH5.5_1  wt_pH5.5_2  wt_pH5.0_1  \\\n",
       "0   lp_0001      423.10      406.42      468.07      465.00      619.47   \n",
       "1   lp_0002      672.64      645.10      681.56      685.90      897.24   \n",
       "2   lp_0004      510.91      559.08      522.82      514.83      499.21   \n",
       "3   lp_0005      358.58      345.58      294.79      299.09      288.61   \n",
       "4   lp_0006      437.95      431.14      357.30      361.36      354.50   \n",
       "\n",
       "   wt_pH5.0_2  \n",
       "0      640.82  \n",
       "1      916.21  \n",
       "2      448.14  \n",
       "3      290.84  \n",
       "4      360.22  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE143834 = pd.read_csv('../data/transcriptomics/GSE143834_Lactobacillus_plantarum_gene_expression_pH_.csv')\n",
    "GSE143834 = GSE143834[['Gene name','Start','End','Strand',\n",
    "                       'con_1','con_2','pH5.0_1','pH5.0_1.1','pH5.5_1','pH5.5_2']]\n",
    "GSE143834 = GSE143834.rename(columns={'con_1':'wt_pH6.2_1', 'con_2':'wt_pH6.2_2', 'pH5.0_1':'wt_pH5.0_1',\n",
    "                                'pH5.0_1.1':'wt_pH5.0_2', 'pH5.5_1':'wt_pH5.5_1', 'pH5.5_2':'wt_pH5.5_2'})\n",
    "IDs = []\n",
    "for i in range(len(GSE143834.index)):\n",
    "    name = list(GSE143834['Gene name'])[i]\n",
    "    if 'lp_' in name:\n",
    "        IDs.append(name)\n",
    "    elif name in list(ref_features['symbol']):\n",
    "        temp = ref_features[ref_features['symbol'] == name ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    elif name in list(gb_features['symbol']):\n",
    "        temp = gb_features[gb_features['symbol'] == name ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    else:\n",
    "        IDs.append(name)\n",
    "GSE143834['locus_tag'] = IDs\n",
    "GSE143834 = GSE143834[['locus_tag','wt_pH6.2_1','wt_pH6.2_2','wt_pH5.5_1','wt_pH5.5_2',\n",
    "                       'wt_pH5.0_1','wt_pH5.0_2']]\n",
    "print(GSE143834.shape)\n",
    "GSE143834.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ec8fabd",
   "metadata": {},
   "source": [
    "## L. plantarum different carbon sources (Convert count to RPKM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "10225a0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3058, 23)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>wt_GLC_rep1</th>\n",
       "      <th>wt_GLC_rep2</th>\n",
       "      <th>wt_PAC1_rep1</th>\n",
       "      <th>wt_PAC1_rep2</th>\n",
       "      <th>wt_PAC2_rep1</th>\n",
       "      <th>wt_PAC2_rep2</th>\n",
       "      <th>wt_FOSPAC1_rep1</th>\n",
       "      <th>wt_FOSPAC1_rep2</th>\n",
       "      <th>wt_FOSPAC2_rep1</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_HMOPAC1_rep1</th>\n",
       "      <th>wt_HMOPAC1_rep2</th>\n",
       "      <th>wt_HMOPAC2_rep1</th>\n",
       "      <th>wt_HMOPAC2_rep2</th>\n",
       "      <th>wt_XG_rep1</th>\n",
       "      <th>wt_XG_rep2</th>\n",
       "      <th>wt_XGPAC1_rep1</th>\n",
       "      <th>wt_XGPAC1_rep2</th>\n",
       "      <th>wt_XGPAC2_rep1</th>\n",
       "      <th>wt_XGPAC2_rep2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>310.016351</td>\n",
       "      <td>378.385267</td>\n",
       "      <td>186.335784</td>\n",
       "      <td>181.704531</td>\n",
       "      <td>174.767584</td>\n",
       "      <td>331.888344</td>\n",
       "      <td>274.911717</td>\n",
       "      <td>307.043525</td>\n",
       "      <td>133.515007</td>\n",
       "      <td>...</td>\n",
       "      <td>502.441837</td>\n",
       "      <td>465.934221</td>\n",
       "      <td>223.878364</td>\n",
       "      <td>190.776541</td>\n",
       "      <td>145.272896</td>\n",
       "      <td>114.711457</td>\n",
       "      <td>106.658059</td>\n",
       "      <td>150.624559</td>\n",
       "      <td>168.328237</td>\n",
       "      <td>192.070155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>669.481193</td>\n",
       "      <td>750.643207</td>\n",
       "      <td>618.965397</td>\n",
       "      <td>640.799551</td>\n",
       "      <td>468.851070</td>\n",
       "      <td>484.447600</td>\n",
       "      <td>459.368611</td>\n",
       "      <td>485.732790</td>\n",
       "      <td>602.754007</td>\n",
       "      <td>...</td>\n",
       "      <td>713.449932</td>\n",
       "      <td>623.604566</td>\n",
       "      <td>609.847872</td>\n",
       "      <td>572.761569</td>\n",
       "      <td>514.850531</td>\n",
       "      <td>449.567228</td>\n",
       "      <td>297.530491</td>\n",
       "      <td>352.205911</td>\n",
       "      <td>553.273363</td>\n",
       "      <td>581.633621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>987.389048</td>\n",
       "      <td>994.441856</td>\n",
       "      <td>289.907038</td>\n",
       "      <td>324.869788</td>\n",
       "      <td>542.896017</td>\n",
       "      <td>391.251387</td>\n",
       "      <td>465.814018</td>\n",
       "      <td>456.112576</td>\n",
       "      <td>390.274635</td>\n",
       "      <td>...</td>\n",
       "      <td>307.528012</td>\n",
       "      <td>351.971933</td>\n",
       "      <td>528.534823</td>\n",
       "      <td>521.880441</td>\n",
       "      <td>269.890260</td>\n",
       "      <td>310.231795</td>\n",
       "      <td>480.214011</td>\n",
       "      <td>470.846952</td>\n",
       "      <td>397.137958</td>\n",
       "      <td>350.071757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>114.401974</td>\n",
       "      <td>106.758145</td>\n",
       "      <td>34.261741</td>\n",
       "      <td>33.250165</td>\n",
       "      <td>62.854717</td>\n",
       "      <td>71.754663</td>\n",
       "      <td>190.020699</td>\n",
       "      <td>190.082872</td>\n",
       "      <td>74.980397</td>\n",
       "      <td>...</td>\n",
       "      <td>72.678780</td>\n",
       "      <td>80.916495</td>\n",
       "      <td>124.571592</td>\n",
       "      <td>74.847756</td>\n",
       "      <td>53.047054</td>\n",
       "      <td>90.030381</td>\n",
       "      <td>126.315493</td>\n",
       "      <td>114.415809</td>\n",
       "      <td>69.334865</td>\n",
       "      <td>85.637680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>87.535495</td>\n",
       "      <td>76.947033</td>\n",
       "      <td>86.446226</td>\n",
       "      <td>82.303207</td>\n",
       "      <td>108.122124</td>\n",
       "      <td>100.719180</td>\n",
       "      <td>395.379766</td>\n",
       "      <td>374.195643</td>\n",
       "      <td>172.343487</td>\n",
       "      <td>...</td>\n",
       "      <td>61.886886</td>\n",
       "      <td>61.782763</td>\n",
       "      <td>102.612308</td>\n",
       "      <td>42.489197</td>\n",
       "      <td>53.565249</td>\n",
       "      <td>114.534706</td>\n",
       "      <td>183.798398</td>\n",
       "      <td>146.368833</td>\n",
       "      <td>60.189621</td>\n",
       "      <td>78.325210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  wt_GLC_rep1  wt_GLC_rep2  wt_PAC1_rep1  wt_PAC1_rep2  \\\n",
       "0   lp_0001   310.016351   378.385267    186.335784    181.704531   \n",
       "1   lp_0002   669.481193   750.643207    618.965397    640.799551   \n",
       "2   lp_0004   987.389048   994.441856    289.907038    324.869788   \n",
       "3   lp_0005   114.401974   106.758145     34.261741     33.250165   \n",
       "4   lp_0006    87.535495    76.947033     86.446226     82.303207   \n",
       "\n",
       "   wt_PAC2_rep1  wt_PAC2_rep2  wt_FOSPAC1_rep1  wt_FOSPAC1_rep2  \\\n",
       "0    174.767584    331.888344       274.911717       307.043525   \n",
       "1    468.851070    484.447600       459.368611       485.732790   \n",
       "2    542.896017    391.251387       465.814018       456.112576   \n",
       "3     62.854717     71.754663       190.020699       190.082872   \n",
       "4    108.122124    100.719180       395.379766       374.195643   \n",
       "\n",
       "   wt_FOSPAC2_rep1  ...  wt_HMOPAC1_rep1  wt_HMOPAC1_rep2  wt_HMOPAC2_rep1  \\\n",
       "0       133.515007  ...       502.441837       465.934221       223.878364   \n",
       "1       602.754007  ...       713.449932       623.604566       609.847872   \n",
       "2       390.274635  ...       307.528012       351.971933       528.534823   \n",
       "3        74.980397  ...        72.678780        80.916495       124.571592   \n",
       "4       172.343487  ...        61.886886        61.782763       102.612308   \n",
       "\n",
       "   wt_HMOPAC2_rep2  wt_XG_rep1  wt_XG_rep2  wt_XGPAC1_rep1  wt_XGPAC1_rep2  \\\n",
       "0       190.776541  145.272896  114.711457      106.658059      150.624559   \n",
       "1       572.761569  514.850531  449.567228      297.530491      352.205911   \n",
       "2       521.880441  269.890260  310.231795      480.214011      470.846952   \n",
       "3        74.847756   53.047054   90.030381      126.315493      114.415809   \n",
       "4        42.489197   53.565249  114.534706      183.798398      146.368833   \n",
       "\n",
       "   wt_XGPAC2_rep1  wt_XGPAC2_rep2  \n",
       "0      168.328237      192.070155  \n",
       "1      553.273363      581.633621  \n",
       "2      397.137958      350.071757  \n",
       "3       69.334865       85.637680  \n",
       "4       60.189621       78.325210  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE160565 = pd.read_csv('../data/transcriptomics/GSE160565_raw_gene_count_matrix.txt',sep='\\t')\n",
    "GSE160565 = GSE160565.dropna()\n",
    "GSE160565 = GSE160565.rename(columns={'Unnamed: 0':'gene'})\n",
    "sample_list = list(GSE160565.columns)[1:]\n",
    "sample_names = [\"LP_GLC_rep1\",\"LP_GLC_rep2\",\"LP_PAC1_rep1\",\"LP_PAC1_rep2\",\"LP_PAC2_rep1\",\"LP_PAC2_rep2\",\n",
    "                \"LP_FOSPAC1_rep1\",\"LP_FOSPAC1_rep2\",\"LP_FOSPAC2_rep1\",\"LP_FOSPAC2_rep2\",\"LP_HMO_rep1\",\n",
    "                \"LP_HMO_rep2\",\"LP_HMOPAC1_rep1\",\"LP_HMOPAC1_rep2\",\"LP_HMOPAC2_rep1\",\"LP_HMOPAC2_rep2\",\n",
    "                \"LP_XG_rep1\",\"LP_XG_rep2\",\"LP_XGPAC1_rep1\",\"LP_XGPAC1_rep2\",\"LP_XGPAC2_rep1\",\"LP_XGPAC2_rep2\"]\n",
    "rename_dict = {sample_list[i]:sample_names[i].replace('LP','wt') for i in range(len(sample_names))}\n",
    "GSE160565 = GSE160565.rename(columns=rename_dict)\n",
    "sample_list = list(GSE160565.columns)[1:]\n",
    "totalReads = {sample: sum(GSE160565[sample]) for sample in sample_list }\n",
    "IDs = []\n",
    "for i in range(len( GSE160565.index )):\n",
    "    gene_id = list(GSE160565['gene'])[i].split('-')[1].strip()\n",
    "    if 'RNA' in list(GSE160565['gene'])[i]:\n",
    "        IDs.append('-')\n",
    "    elif ( gene_id in list(ref_features['locus_tag']) ) or ( gene_id in list(gb_features['locus_tag']) ):\n",
    "        IDs.append( gene_id )\n",
    "    else:\n",
    "        IDs.append('-')\n",
    "GSE160565['locus_tag'] = IDs\n",
    "GSE160565 = GSE160565[GSE160565['locus_tag'] != '-']\n",
    "lengths=[]\n",
    "for i in range(len( GSE160565.index )):\n",
    "    gene_id = list(GSE160565['locus_tag'])[i]\n",
    "    if gene_id in list(ref_features['locus_tag']):\n",
    "        temp = ref_features[ref_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )\n",
    "    else:\n",
    "        temp = gb_features[gb_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )       \n",
    "GSE160565['length'] = lengths\n",
    "\n",
    "for sample in sample_list:\n",
    "    temp_rpkm = [ get_rpkm( list(GSE160565[sample])[i],\n",
    "                    list(GSE160565['length'])[i], totalReads[sample]) for i in range(len(GSE160565.index))]\n",
    "    GSE160565[sample] = temp_rpkm\n",
    "GSE160565 = GSE160565[['locus_tag']+sample_list]\n",
    "print(GSE160565.shape)\n",
    "GSE160565.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52cada60",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ed85a99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# geo_merged = reduce(lambda  left,right: pd.merge(left,right,on=['locus_tag'],\n",
    "#                             how='inner'), [GSE143834, GSE124050, GSE115448, GSE160565 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "52e19a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# geo_merged.to_csv('../data/transcriptomics/merged_rpkm.csv', index = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbbf4d58",
   "metadata": {},
   "source": [
    "# Log10 transformation and center to reference condition(wt_pH6.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "fc3f6297",
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_merged = pd.read_csv('../data/transcriptomics/merged_rpkm.csv')\n",
    "log_rpkm = geo_merged.copy()\n",
    "sample_list = list(geo_merged.columns)[1:]\n",
    "for sample in sample_list:\n",
    "    temp_log10 = [ np.log10(  list(geo_merged[sample])[i] + 1  ) for i in range(len(log_rpkm.index)) ]\n",
    "    log_rpkm[ sample ] = temp_log10\n",
    "log_rpkm = log_rpkm.set_index(['locus_tag'])\n",
    "# log_rpkm.to_csv('../data/transcriptomics/log_rpkm.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e53651f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_pH6.2_1</th>\n",
       "      <th>wt_pH6.2_2</th>\n",
       "      <th>wt_pH5.5_1</th>\n",
       "      <th>wt_pH5.5_2</th>\n",
       "      <th>wt_pH5.0_1</th>\n",
       "      <th>wt_pH5.0_2</th>\n",
       "      <th>wt_3OC_1m1b</th>\n",
       "      <th>wt_3OC_1m4b</th>\n",
       "      <th>wt_3OC_1m7b</th>\n",
       "      <th>wt_3OC_1p1b</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_HMOPAC1_rep1</th>\n",
       "      <th>wt_HMOPAC1_rep2</th>\n",
       "      <th>wt_HMOPAC2_rep1</th>\n",
       "      <th>wt_HMOPAC2_rep2</th>\n",
       "      <th>wt_XG_rep1</th>\n",
       "      <th>wt_XG_rep2</th>\n",
       "      <th>wt_XGPAC1_rep1</th>\n",
       "      <th>wt_XGPAC1_rep2</th>\n",
       "      <th>wt_XGPAC2_rep1</th>\n",
       "      <th>wt_XGPAC2_rep2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lp_0001</th>\n",
       "      <td>2.627468</td>\n",
       "      <td>2.610042</td>\n",
       "      <td>2.671238</td>\n",
       "      <td>2.668386</td>\n",
       "      <td>2.792721</td>\n",
       "      <td>2.807413</td>\n",
       "      <td>2.615262</td>\n",
       "      <td>2.743785</td>\n",
       "      <td>2.398585</td>\n",
       "      <td>2.323148</td>\n",
       "      <td>...</td>\n",
       "      <td>2.701949</td>\n",
       "      <td>2.669256</td>\n",
       "      <td>2.351948</td>\n",
       "      <td>2.282795</td>\n",
       "      <td>2.165164</td>\n",
       "      <td>2.063376</td>\n",
       "      <td>2.032047</td>\n",
       "      <td>2.180770</td>\n",
       "      <td>2.228729</td>\n",
       "      <td>2.285715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0002</th>\n",
       "      <td>2.828428</td>\n",
       "      <td>2.810300</td>\n",
       "      <td>2.834141</td>\n",
       "      <td>2.836894</td>\n",
       "      <td>2.953392</td>\n",
       "      <td>2.962469</td>\n",
       "      <td>2.830527</td>\n",
       "      <td>2.863596</td>\n",
       "      <td>2.669473</td>\n",
       "      <td>2.545511</td>\n",
       "      <td>...</td>\n",
       "      <td>2.853972</td>\n",
       "      <td>2.795605</td>\n",
       "      <td>2.785933</td>\n",
       "      <td>2.758731</td>\n",
       "      <td>2.712524</td>\n",
       "      <td>2.653760</td>\n",
       "      <td>2.474989</td>\n",
       "      <td>2.548028</td>\n",
       "      <td>2.743724</td>\n",
       "      <td>2.765396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0004</th>\n",
       "      <td>2.709194</td>\n",
       "      <td>2.748250</td>\n",
       "      <td>2.719182</td>\n",
       "      <td>2.712507</td>\n",
       "      <td>2.699152</td>\n",
       "      <td>2.652382</td>\n",
       "      <td>2.577278</td>\n",
       "      <td>2.554357</td>\n",
       "      <td>2.788437</td>\n",
       "      <td>2.437963</td>\n",
       "      <td>...</td>\n",
       "      <td>2.489295</td>\n",
       "      <td>2.547740</td>\n",
       "      <td>2.723895</td>\n",
       "      <td>2.718402</td>\n",
       "      <td>2.432793</td>\n",
       "      <td>2.493084</td>\n",
       "      <td>2.682338</td>\n",
       "      <td>2.673801</td>\n",
       "      <td>2.600034</td>\n",
       "      <td>2.545396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0005</th>\n",
       "      <td>2.555796</td>\n",
       "      <td>2.539803</td>\n",
       "      <td>2.470983</td>\n",
       "      <td>2.477252</td>\n",
       "      <td>2.461814</td>\n",
       "      <td>2.465145</td>\n",
       "      <td>2.424988</td>\n",
       "      <td>2.412567</td>\n",
       "      <td>2.557048</td>\n",
       "      <td>2.431580</td>\n",
       "      <td>...</td>\n",
       "      <td>1.867342</td>\n",
       "      <td>1.913371</td>\n",
       "      <td>2.098891</td>\n",
       "      <td>1.879943</td>\n",
       "      <td>1.732772</td>\n",
       "      <td>1.959186</td>\n",
       "      <td>2.104881</td>\n",
       "      <td>2.062265</td>\n",
       "      <td>1.847171</td>\n",
       "      <td>1.937707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0006</th>\n",
       "      <td>2.642415</td>\n",
       "      <td>2.635624</td>\n",
       "      <td>2.554247</td>\n",
       "      <td>2.559140</td>\n",
       "      <td>2.550840</td>\n",
       "      <td>2.557772</td>\n",
       "      <td>2.637792</td>\n",
       "      <td>2.645050</td>\n",
       "      <td>2.582206</td>\n",
       "      <td>2.656157</td>\n",
       "      <td>...</td>\n",
       "      <td>1.798560</td>\n",
       "      <td>1.797840</td>\n",
       "      <td>2.015411</td>\n",
       "      <td>1.638381</td>\n",
       "      <td>1.736916</td>\n",
       "      <td>2.062712</td>\n",
       "      <td>2.266698</td>\n",
       "      <td>2.168406</td>\n",
       "      <td>1.786678</td>\n",
       "      <td>1.899411</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           wt_pH6.2_1  wt_pH6.2_2  wt_pH5.5_1  wt_pH5.5_2  wt_pH5.0_1  \\\n",
       "locus_tag                                                               \n",
       "lp_0001      2.627468    2.610042    2.671238    2.668386    2.792721   \n",
       "lp_0002      2.828428    2.810300    2.834141    2.836894    2.953392   \n",
       "lp_0004      2.709194    2.748250    2.719182    2.712507    2.699152   \n",
       "lp_0005      2.555796    2.539803    2.470983    2.477252    2.461814   \n",
       "lp_0006      2.642415    2.635624    2.554247    2.559140    2.550840   \n",
       "\n",
       "           wt_pH5.0_2  wt_3OC_1m1b  wt_3OC_1m4b  wt_3OC_1m7b  wt_3OC_1p1b  \\\n",
       "locus_tag                                                                   \n",
       "lp_0001      2.807413     2.615262     2.743785     2.398585     2.323148   \n",
       "lp_0002      2.962469     2.830527     2.863596     2.669473     2.545511   \n",
       "lp_0004      2.652382     2.577278     2.554357     2.788437     2.437963   \n",
       "lp_0005      2.465145     2.424988     2.412567     2.557048     2.431580   \n",
       "lp_0006      2.557772     2.637792     2.645050     2.582206     2.656157   \n",
       "\n",
       "           ...  wt_HMOPAC1_rep1  wt_HMOPAC1_rep2  wt_HMOPAC2_rep1  \\\n",
       "locus_tag  ...                                                      \n",
       "lp_0001    ...         2.701949         2.669256         2.351948   \n",
       "lp_0002    ...         2.853972         2.795605         2.785933   \n",
       "lp_0004    ...         2.489295         2.547740         2.723895   \n",
       "lp_0005    ...         1.867342         1.913371         2.098891   \n",
       "lp_0006    ...         1.798560         1.797840         2.015411   \n",
       "\n",
       "           wt_HMOPAC2_rep2  wt_XG_rep1  wt_XG_rep2  wt_XGPAC1_rep1  \\\n",
       "locus_tag                                                            \n",
       "lp_0001           2.282795    2.165164    2.063376        2.032047   \n",
       "lp_0002           2.758731    2.712524    2.653760        2.474989   \n",
       "lp_0004           2.718402    2.432793    2.493084        2.682338   \n",
       "lp_0005           1.879943    1.732772    1.959186        2.104881   \n",
       "lp_0006           1.638381    1.736916    2.062712        2.266698   \n",
       "\n",
       "           wt_XGPAC1_rep2  wt_XGPAC2_rep1  wt_XGPAC2_rep2  \n",
       "locus_tag                                                  \n",
       "lp_0001          2.180770        2.228729        2.285715  \n",
       "lp_0002          2.548028        2.743724        2.765396  \n",
       "lp_0004          2.673801        2.600034        2.545396  \n",
       "lp_0005          2.062265        1.847171        1.937707  \n",
       "lp_0006          2.168406        1.786678        1.899411  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_rpkm = pd.read_csv('../data/transcriptomics/log_rpkm.csv',index_col=0)\n",
    "log_rpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f504fdef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_pH6.2_1</th>\n",
       "      <th>wt_pH6.2_2</th>\n",
       "      <th>wt_pH5.5_1</th>\n",
       "      <th>wt_pH5.5_2</th>\n",
       "      <th>wt_pH5.0_1</th>\n",
       "      <th>wt_pH5.0_2</th>\n",
       "      <th>wt_3OC_1m1b</th>\n",
       "      <th>wt_3OC_1m4b</th>\n",
       "      <th>wt_3OC_1m7b</th>\n",
       "      <th>wt_3OC_1p1b</th>\n",
       "      <th>...</th>\n",
       "      <th>wt_HMOPAC1_rep1</th>\n",
       "      <th>wt_HMOPAC1_rep2</th>\n",
       "      <th>wt_HMOPAC2_rep1</th>\n",
       "      <th>wt_HMOPAC2_rep2</th>\n",
       "      <th>wt_XG_rep1</th>\n",
       "      <th>wt_XG_rep2</th>\n",
       "      <th>wt_XGPAC1_rep1</th>\n",
       "      <th>wt_XGPAC1_rep2</th>\n",
       "      <th>wt_XGPAC2_rep1</th>\n",
       "      <th>wt_XGPAC2_rep2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lp_0001</th>\n",
       "      <td>0.008713</td>\n",
       "      <td>-0.008713</td>\n",
       "      <td>0.052482</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>0.173965</td>\n",
       "      <td>0.188658</td>\n",
       "      <td>-0.003493</td>\n",
       "      <td>0.125030</td>\n",
       "      <td>-0.220170</td>\n",
       "      <td>-0.295607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.083194</td>\n",
       "      <td>0.050500</td>\n",
       "      <td>-0.266808</td>\n",
       "      <td>-0.335960</td>\n",
       "      <td>-0.453591</td>\n",
       "      <td>-0.555379</td>\n",
       "      <td>-0.586709</td>\n",
       "      <td>-0.437986</td>\n",
       "      <td>-0.390026</td>\n",
       "      <td>-0.333040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0002</th>\n",
       "      <td>0.009064</td>\n",
       "      <td>-0.009064</td>\n",
       "      <td>0.014777</td>\n",
       "      <td>0.017530</td>\n",
       "      <td>0.134029</td>\n",
       "      <td>0.143105</td>\n",
       "      <td>0.011163</td>\n",
       "      <td>0.044232</td>\n",
       "      <td>-0.149891</td>\n",
       "      <td>-0.273853</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034608</td>\n",
       "      <td>-0.023759</td>\n",
       "      <td>-0.033431</td>\n",
       "      <td>-0.060632</td>\n",
       "      <td>-0.106840</td>\n",
       "      <td>-0.165604</td>\n",
       "      <td>-0.344375</td>\n",
       "      <td>-0.271336</td>\n",
       "      <td>-0.075640</td>\n",
       "      <td>-0.053968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0004</th>\n",
       "      <td>-0.019528</td>\n",
       "      <td>0.019528</td>\n",
       "      <td>-0.009540</td>\n",
       "      <td>-0.016215</td>\n",
       "      <td>-0.029569</td>\n",
       "      <td>-0.076340</td>\n",
       "      <td>-0.151444</td>\n",
       "      <td>-0.174365</td>\n",
       "      <td>0.059716</td>\n",
       "      <td>-0.290759</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.239427</td>\n",
       "      <td>-0.180982</td>\n",
       "      <td>-0.004827</td>\n",
       "      <td>-0.010319</td>\n",
       "      <td>-0.295928</td>\n",
       "      <td>-0.235638</td>\n",
       "      <td>-0.046384</td>\n",
       "      <td>-0.054921</td>\n",
       "      <td>-0.128688</td>\n",
       "      <td>-0.183326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0005</th>\n",
       "      <td>0.007996</td>\n",
       "      <td>-0.007996</td>\n",
       "      <td>-0.076816</td>\n",
       "      <td>-0.070548</td>\n",
       "      <td>-0.085986</td>\n",
       "      <td>-0.082655</td>\n",
       "      <td>-0.122812</td>\n",
       "      <td>-0.135233</td>\n",
       "      <td>0.009248</td>\n",
       "      <td>-0.116220</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.680457</td>\n",
       "      <td>-0.634428</td>\n",
       "      <td>-0.448908</td>\n",
       "      <td>-0.667857</td>\n",
       "      <td>-0.815027</td>\n",
       "      <td>-0.588613</td>\n",
       "      <td>-0.442918</td>\n",
       "      <td>-0.485534</td>\n",
       "      <td>-0.700629</td>\n",
       "      <td>-0.610093</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lp_0006</th>\n",
       "      <td>0.003395</td>\n",
       "      <td>-0.003395</td>\n",
       "      <td>-0.084773</td>\n",
       "      <td>-0.079880</td>\n",
       "      <td>-0.088180</td>\n",
       "      <td>-0.081248</td>\n",
       "      <td>-0.001228</td>\n",
       "      <td>0.006030</td>\n",
       "      <td>-0.056814</td>\n",
       "      <td>0.017137</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.840460</td>\n",
       "      <td>-0.841179</td>\n",
       "      <td>-0.623608</td>\n",
       "      <td>-1.000638</td>\n",
       "      <td>-0.902104</td>\n",
       "      <td>-0.576307</td>\n",
       "      <td>-0.372322</td>\n",
       "      <td>-0.470614</td>\n",
       "      <td>-0.852342</td>\n",
       "      <td>-0.739609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           wt_pH6.2_1  wt_pH6.2_2  wt_pH5.5_1  wt_pH5.5_2  wt_pH5.0_1  \\\n",
       "locus_tag                                                               \n",
       "lp_0001      0.008713   -0.008713    0.052482    0.049631    0.173965   \n",
       "lp_0002      0.009064   -0.009064    0.014777    0.017530    0.134029   \n",
       "lp_0004     -0.019528    0.019528   -0.009540   -0.016215   -0.029569   \n",
       "lp_0005      0.007996   -0.007996   -0.076816   -0.070548   -0.085986   \n",
       "lp_0006      0.003395   -0.003395   -0.084773   -0.079880   -0.088180   \n",
       "\n",
       "           wt_pH5.0_2  wt_3OC_1m1b  wt_3OC_1m4b  wt_3OC_1m7b  wt_3OC_1p1b  \\\n",
       "locus_tag                                                                   \n",
       "lp_0001      0.188658    -0.003493     0.125030    -0.220170    -0.295607   \n",
       "lp_0002      0.143105     0.011163     0.044232    -0.149891    -0.273853   \n",
       "lp_0004     -0.076340    -0.151444    -0.174365     0.059716    -0.290759   \n",
       "lp_0005     -0.082655    -0.122812    -0.135233     0.009248    -0.116220   \n",
       "lp_0006     -0.081248    -0.001228     0.006030    -0.056814     0.017137   \n",
       "\n",
       "           ...  wt_HMOPAC1_rep1  wt_HMOPAC1_rep2  wt_HMOPAC2_rep1  \\\n",
       "locus_tag  ...                                                      \n",
       "lp_0001    ...         0.083194         0.050500        -0.266808   \n",
       "lp_0002    ...         0.034608        -0.023759        -0.033431   \n",
       "lp_0004    ...        -0.239427        -0.180982        -0.004827   \n",
       "lp_0005    ...        -0.680457        -0.634428        -0.448908   \n",
       "lp_0006    ...        -0.840460        -0.841179        -0.623608   \n",
       "\n",
       "           wt_HMOPAC2_rep2  wt_XG_rep1  wt_XG_rep2  wt_XGPAC1_rep1  \\\n",
       "locus_tag                                                            \n",
       "lp_0001          -0.335960   -0.453591   -0.555379       -0.586709   \n",
       "lp_0002          -0.060632   -0.106840   -0.165604       -0.344375   \n",
       "lp_0004          -0.010319   -0.295928   -0.235638       -0.046384   \n",
       "lp_0005          -0.667857   -0.815027   -0.588613       -0.442918   \n",
       "lp_0006          -1.000638   -0.902104   -0.576307       -0.372322   \n",
       "\n",
       "           wt_XGPAC1_rep2  wt_XGPAC2_rep1  wt_XGPAC2_rep2  \n",
       "locus_tag                                                  \n",
       "lp_0001         -0.437986       -0.390026       -0.333040  \n",
       "lp_0002         -0.271336       -0.075640       -0.053968  \n",
       "lp_0004         -0.054921       -0.128688       -0.183326  \n",
       "lp_0005         -0.485534       -0.700629       -0.610093  \n",
       "lp_0006         -0.470614       -0.852342       -0.739609  \n",
       "\n",
       "[5 rows x 100 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_rpkm_norm = log_rpkm.sub(log_rpkm[['wt_pH6.2_1','wt_pH6.2_2']].mean(axis=1),axis=0)\n",
    "# log_rpkm_norm.to_csv('../data/transcriptomics/log_rpkm_norm.csv')\n",
    "log_rpkm_norm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d08e9974",
   "metadata": {},
   "source": [
    "# Regulons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b9636d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_regprecise(file):\n",
    "    regulon = {}\n",
    "    ffile = open(file, \"rt\")\n",
    "    lines = ffile.readlines()\n",
    "    ffile.close()\n",
    "    for line in lines:\n",
    "        line = line.strip()\n",
    "        if len(line)<1:\n",
    "            continue\n",
    "        elif '#' in line:\n",
    "            RF = ((line.split('-')[1]).split(':')[0]).strip()\n",
    "            regulon[ RF ] = []\n",
    "        else:\n",
    "            g = str(line.split('\\t')[2]).strip()\n",
    "            regulon[RF].append(g)\n",
    "    return regulon "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157046c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# regulons={}\n",
    "# regulons['Lbrevis'] = read_regprecise('../data/regulon/Regulons_Lbrevis_ATCC_367.txt')\n",
    "# regulons['Lplantarum'] = read_regprecise('../data/regulon/Regulons_Lplantarum_WCFS1.txt')\n",
    "# regulons['Lrhamnosus'] = read_regprecise('../data/regulon/Regulons_Lrhamnosus_GG.txt')\n",
    "# dump_pickle(regulons, '../data/regulon/regulons.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7966a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "regulons = load_pickle('../data/regulon/regulons.pkl')\n",
    "print(regulons['Lplantarum'].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d3840a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
