{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "90da34b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os \n",
    "from functions import *\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be11eda1",
   "metadata": {},
   "source": [
    "# Gene-locustags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9d46d93d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_features = pd.read_csv('../data/genome/GCF_000203855.3_ASM20385v3_feature_table.txt',sep='\\t')\n",
    "ref_features = ( ref_features[ref_features['# feature']=='gene']).reset_index().drop(['index'],axis=1)\n",
    "ref_features = ref_features[['start','end','strand','feature_interval_length','symbol','attributes']]\n",
    "ref_features = ( ref_features.dropna(subset=['attributes']) ).reset_index().drop(['index'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46639b25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>symbol</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1368</td>\n",
       "      <td>+</td>\n",
       "      <td>1368</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>lp_0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1546</td>\n",
       "      <td>2685</td>\n",
       "      <td>+</td>\n",
       "      <td>1140</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>lp_0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3210</td>\n",
       "      <td>3443</td>\n",
       "      <td>+</td>\n",
       "      <td>234</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>lp_0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3444</td>\n",
       "      <td>4568</td>\n",
       "      <td>+</td>\n",
       "      <td>1125</td>\n",
       "      <td>recF</td>\n",
       "      <td>lp_0005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4565</td>\n",
       "      <td>6511</td>\n",
       "      <td>+</td>\n",
       "      <td>1947</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>lp_0006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end strand  feature_interval_length symbol locus_tag\n",
       "0      1  1368      +                     1368   dnaA   lp_0001\n",
       "1   1546  2685      +                     1140   dnaN   lp_0002\n",
       "2   3210  3443      +                      234   yaaA   lp_0004\n",
       "3   3444  4568      +                     1125   recF   lp_0005\n",
       "4   4565  6511      +                     1947   gyrB   lp_0006"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "locus_tag=[]\n",
    "for i in range(len(ref_features.index)):\n",
    "    att = list(ref_features['attributes'])[i]\n",
    "    if 'old_locus_tag=' in str( att ):\n",
    "        locus_tag.append( str(att.split('=')[1]).strip()  )\n",
    "    else:\n",
    "        locus_tag.append('-')\n",
    "ref_features['locus_tag']=locus_tag\n",
    "ref_features = ref_features.drop(['attributes'], axis=1)\n",
    "ref_features['symbol'] = ref_features['symbol'].fillna(ref_features['locus_tag'])\n",
    "ref_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fe7ea16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>symbol</th>\n",
       "      <th>locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1368</td>\n",
       "      <td>+</td>\n",
       "      <td>1368</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>lp_0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1546</td>\n",
       "      <td>2685</td>\n",
       "      <td>+</td>\n",
       "      <td>1140</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>lp_0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3210</td>\n",
       "      <td>3443</td>\n",
       "      <td>+</td>\n",
       "      <td>234</td>\n",
       "      <td>lp_0004</td>\n",
       "      <td>lp_0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3444</td>\n",
       "      <td>4568</td>\n",
       "      <td>+</td>\n",
       "      <td>1125</td>\n",
       "      <td>recF</td>\n",
       "      <td>lp_0005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4565</td>\n",
       "      <td>6511</td>\n",
       "      <td>+</td>\n",
       "      <td>1947</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>lp_0006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end strand  feature_interval_length   symbol locus_tag\n",
       "0      1  1368      +                     1368     dnaA   lp_0001\n",
       "1   1546  2685      +                     1140     dnaN   lp_0002\n",
       "2   3210  3443      +                      234  lp_0004   lp_0004\n",
       "3   3444  4568      +                     1125     recF   lp_0005\n",
       "4   4565  6511      +                     1947     gyrB   lp_0006"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb_features = pd.read_csv('../data/genome/GCA_000203855.3_ASM20385v3_feature_table.txt',sep='\\t')\n",
    "gb_features = (gb_features[gb_features['# feature']=='gene']).reset_index().drop(['index'],axis=1)\n",
    "gb_features = gb_features[['start','end','strand','feature_interval_length','symbol','locus_tag']]\n",
    "gb_features = ( gb_features.dropna(subset=['locus_tag']) ).reset_index().drop(['index'],axis=1)\n",
    "gb_features['symbol'] = gb_features['symbol'].fillna(gb_features['locus_tag'])\n",
    "gb_features.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "113d7f59",
   "metadata": {},
   "source": [
    "# Lactobacillus plantarum cauh2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2456555",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_clean_fasta( in_path, out_path ):\n",
    "    temp_records = SeqIO.parse( in_path , \"fasta\")\n",
    "    new_records = []\n",
    "    for seq_record in temp_records:\n",
    "        description = str(seq_record.description)\n",
    "        if 'locus_tag=' in description:\n",
    "            ID = str(seq_record.description).split('locus_tag=')[1].split('] [')[0].strip()\n",
    "        elif 'gene=' in description:\n",
    "            ID = str(seq_record.description).split('gene=')[1].split('] [')[0].strip()\n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "        seq = str(seq_record.seq)\n",
    "        new_records.append( SeqRecord(Seq(seq), id = ID, name=\"\",description=\"\") )\n",
    "    SeqIO.write(new_records, out_path ,\"fasta\")\n",
    "    print( len(new_records) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4169a7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save_clean_fasta( \"../data/genome/GCA_000203855.3_ASM20385v3_cds_from_genomic.fna\",\n",
    "#                  \"../data/genome/wcfs1_cds.fna\")\n",
    "# save_clean_fasta( \"../data/genome/GCA_001617525.1_ASM161752v1_cds_from_genomic.fna\",\n",
    "#                  \"../data/genome/cauh2_cds.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2100557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cmd = 'blastn -query ../data/genome/cauh2_cds.fna -subject ../data/genome/wcfs1_cds.fna -evalue 1e-6 -outfmt 6'\n",
    "# cmd += ' -max_target_seqs 5 -out ../data/genome/cauh2vswcfs1.txt'\n",
    "# os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "637fdb9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_table = pd.read_csv('../data/genome/cauh2vswcfs1.txt',sep='\\t',header=None)\n",
    "blast_table = blast_table.rename({0:'cauh2_id',1:'wcfs1_id',10:'evalue',11:'bitscore'},axis=1)\n",
    "cauh2_genes = list(np.unique(blast_table['cauh2_id']))\n",
    "print(len(cauh2_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d68cab9",
   "metadata": {},
   "source": [
    "# Combine transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2300b0",
   "metadata": {},
   "source": [
    "## L. plantarum treated with 3OC12 at three time points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4dfaf954",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3101, 37)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>1m1b_S4</th>\n",
       "      <th>1m4b_S10</th>\n",
       "      <th>1m7b_S16</th>\n",
       "      <th>1p1b_S1</th>\n",
       "      <th>1p4b_S7</th>\n",
       "      <th>1p7b_S13</th>\n",
       "      <th>2m1b_S5</th>\n",
       "      <th>2m4b_S11</th>\n",
       "      <th>2m7b_S17</th>\n",
       "      <th>...</th>\n",
       "      <th>2m7_S17</th>\n",
       "      <th>2p1_S2</th>\n",
       "      <th>2p4_S8</th>\n",
       "      <th>2p7_S14</th>\n",
       "      <th>3m1_S6</th>\n",
       "      <th>3m4_S12</th>\n",
       "      <th>3m7_S18</th>\n",
       "      <th>3p1_S3</th>\n",
       "      <th>3p4_S9</th>\n",
       "      <th>3p7_S15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>1270</td>\n",
       "      <td>1928</td>\n",
       "      <td>839</td>\n",
       "      <td>529</td>\n",
       "      <td>1635</td>\n",
       "      <td>1081</td>\n",
       "      <td>1363</td>\n",
       "      <td>1328</td>\n",
       "      <td>1037</td>\n",
       "      <td>...</td>\n",
       "      <td>1014</td>\n",
       "      <td>641</td>\n",
       "      <td>1728</td>\n",
       "      <td>1059</td>\n",
       "      <td>1078</td>\n",
       "      <td>2357</td>\n",
       "      <td>962</td>\n",
       "      <td>530</td>\n",
       "      <td>1510</td>\n",
       "      <td>1079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>1739</td>\n",
       "      <td>2118</td>\n",
       "      <td>1307</td>\n",
       "      <td>737</td>\n",
       "      <td>1843</td>\n",
       "      <td>1614</td>\n",
       "      <td>1707</td>\n",
       "      <td>1529</td>\n",
       "      <td>1422</td>\n",
       "      <td>...</td>\n",
       "      <td>1274</td>\n",
       "      <td>899</td>\n",
       "      <td>1761</td>\n",
       "      <td>1498</td>\n",
       "      <td>1515</td>\n",
       "      <td>8305</td>\n",
       "      <td>1359</td>\n",
       "      <td>754</td>\n",
       "      <td>1517</td>\n",
       "      <td>1499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>199</td>\n",
       "      <td>213</td>\n",
       "      <td>353</td>\n",
       "      <td>118</td>\n",
       "      <td>314</td>\n",
       "      <td>383</td>\n",
       "      <td>219</td>\n",
       "      <td>194</td>\n",
       "      <td>267</td>\n",
       "      <td>...</td>\n",
       "      <td>300</td>\n",
       "      <td>237</td>\n",
       "      <td>231</td>\n",
       "      <td>290</td>\n",
       "      <td>185</td>\n",
       "      <td>79</td>\n",
       "      <td>202</td>\n",
       "      <td>210</td>\n",
       "      <td>250</td>\n",
       "      <td>367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>673</td>\n",
       "      <td>738</td>\n",
       "      <td>995</td>\n",
       "      <td>559</td>\n",
       "      <td>699</td>\n",
       "      <td>1006</td>\n",
       "      <td>804</td>\n",
       "      <td>672</td>\n",
       "      <td>848</td>\n",
       "      <td>...</td>\n",
       "      <td>881</td>\n",
       "      <td>839</td>\n",
       "      <td>613</td>\n",
       "      <td>854</td>\n",
       "      <td>618</td>\n",
       "      <td>1096</td>\n",
       "      <td>691</td>\n",
       "      <td>705</td>\n",
       "      <td>673</td>\n",
       "      <td>1031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>1904</td>\n",
       "      <td>2185</td>\n",
       "      <td>1825</td>\n",
       "      <td>1625</td>\n",
       "      <td>1836</td>\n",
       "      <td>1734</td>\n",
       "      <td>2047</td>\n",
       "      <td>2071</td>\n",
       "      <td>1772</td>\n",
       "      <td>...</td>\n",
       "      <td>1840</td>\n",
       "      <td>2184</td>\n",
       "      <td>1801</td>\n",
       "      <td>1594</td>\n",
       "      <td>1810</td>\n",
       "      <td>1147</td>\n",
       "      <td>1767</td>\n",
       "      <td>1854</td>\n",
       "      <td>1835</td>\n",
       "      <td>1776</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  1m1b_S4  1m4b_S10  1m7b_S16  1p1b_S1  1p4b_S7  1p7b_S13  2m1b_S5  \\\n",
       "0   lp_0001     1270      1928       839      529     1635      1081     1363   \n",
       "1   lp_0002     1739      2118      1307      737     1843      1614     1707   \n",
       "2   lp_0004      199       213       353      118      314       383      219   \n",
       "3   lp_0005      673       738       995      559      699      1006      804   \n",
       "4   lp_0006     1904      2185      1825     1625     1836      1734     2047   \n",
       "\n",
       "   2m4b_S11  2m7b_S17  ...  2m7_S17  2p1_S2  2p4_S8  2p7_S14  3m1_S6  3m4_S12  \\\n",
       "0      1328      1037  ...     1014     641    1728     1059    1078     2357   \n",
       "1      1529      1422  ...     1274     899    1761     1498    1515     8305   \n",
       "2       194       267  ...      300     237     231      290     185       79   \n",
       "3       672       848  ...      881     839     613      854     618     1096   \n",
       "4      2071      1772  ...     1840    2184    1801     1594    1810     1147   \n",
       "\n",
       "   3m7_S18  3p1_S3  3p4_S9  3p7_S15  \n",
       "0      962     530    1510     1079  \n",
       "1     1359     754    1517     1499  \n",
       "2      202     210     250      367  \n",
       "3      691     705     673     1031  \n",
       "4     1767    1854    1835     1776  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE124050_rpkm_0_ADTV7 = pd.read_csv('../data/transcriptomics/GSE124050_rpkm_0_ADTV7.txt',sep='\\t')\n",
    "GSE124050_rpkm_0_AYR0M = pd.read_csv('../data/transcriptomics/GSE124050_rpkm_0_AYR0M.txt',sep='\\t')\n",
    "GSE124050 = GSE124050_rpkm_0_ADTV7.merge( GSE124050_rpkm_0_AYR0M, on=['gene'],how='inner')\n",
    "GSE124050 = GSE124050.rename(columns={'gene':'locus_tag'})\n",
    "sample_list = list(GSE124050.columns)[1:]\n",
    "rename_dict = {x: x.split('.')[0].strip() for x in sample_list}\n",
    "GSE124050 = GSE124050.rename(columns=rename_dict)\n",
    "print(GSE124050.shape)\n",
    "GSE124050.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16e05a3",
   "metadata": {},
   "source": [
    "## L. plantarum strains from nine different habitats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0200fc2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3108, 37)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>WCFS1_MRS</th>\n",
       "      <th>LB16_MRS</th>\n",
       "      <th>TJ22</th>\n",
       "      <th>TJ20</th>\n",
       "      <th>TJ19</th>\n",
       "      <th>WFH19</th>\n",
       "      <th>WFH20</th>\n",
       "      <th>WFH1W</th>\n",
       "      <th>PJ16</th>\n",
       "      <th>...</th>\n",
       "      <th>CB5</th>\n",
       "      <th>FEV11_MRS</th>\n",
       "      <th>FEO3_MRS</th>\n",
       "      <th>DE12_MRS</th>\n",
       "      <th>CB5_MRS</th>\n",
       "      <th>OE1_MRS</th>\n",
       "      <th>WFH19_MRS</th>\n",
       "      <th>TJ22_MRS</th>\n",
       "      <th>PJ16_MRS</th>\n",
       "      <th>BEE1ST_MRS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>repA</td>\n",
       "      <td>190.5490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.282865</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>orf2</td>\n",
       "      <td>59.9286</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>246.269</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>33.7182</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.90913</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>orf3</td>\n",
       "      <td>116.1010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>repB</td>\n",
       "      <td>357.2480</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.361062</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>orf2</td>\n",
       "      <td>203.2870</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  WCFS1_MRS  LB16_MRS  TJ22  TJ20      TJ19     WFH19  WFH20  \\\n",
       "0      repA   190.5490       0.0   0.0   0.0  0.000000  0.282865    0.0   \n",
       "1      orf2    59.9286       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "2      orf3   116.1010       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "3      repB   357.2480       0.0   0.0   0.0  0.361062  0.000000    0.0   \n",
       "4      orf2   203.2870       0.0   0.0   0.0  0.000000  0.000000    0.0   \n",
       "\n",
       "     WFH1W  PJ16  ...      CB5  FEV11_MRS  FEO3_MRS  DE12_MRS  CB5_MRS  \\\n",
       "0    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "1  246.269   0.0  ...  33.7182        0.0       0.0       0.0  3.90913   \n",
       "2    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "3    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "4    0.000   0.0  ...   0.0000        0.0       0.0       0.0  0.00000   \n",
       "\n",
       "   OE1_MRS  WFH19_MRS  TJ22_MRS  PJ16_MRS  BEE1ST_MRS  \n",
       "0      0.0        0.0       0.0       0.0         0.0  \n",
       "1      0.0        0.0       0.0       0.0         0.0  \n",
       "2      0.0        0.0       0.0       0.0         0.0  \n",
       "3      0.0        0.0       0.0       0.0         0.0  \n",
       "4      0.0        0.0       0.0       0.0         0.0  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE115448 = pd.read_csv('../data/transcriptomics/GSE115448_RPKMvalues.txt',sep='\\t')\n",
    "sample_list = list(GSE115448.columns)[2:]\n",
    "IDs = []\n",
    "for i in range(len(GSE115448.index)):\n",
    "    if 'lp_' in list(GSE115448['GeneSymbol'])[i]:\n",
    "        IDs.append( list(GSE115448['GeneSymbol'])[i].split(':')[1].strip() )\n",
    "    elif 'lp_' in list(GSE115448['GeneName'])[i]:\n",
    "        IDs.append( list(GSE115448['GeneName'])[i].strip() )\n",
    "    elif list(GSE115448['GeneName'])[i] in list(ref_features['symbol']):\n",
    "        temp = ref_features[ref_features['symbol'] == list(GSE115448['GeneName'])[i] ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    elif list(GSE115448['GeneName'])[i] in list(gb_features['symbol']):\n",
    "        temp = gb_features[gb_features['symbol'] == list(GSE115448['GeneName'])[i] ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    else:\n",
    "        IDs.append(list(GSE115448['GeneName'])[i])\n",
    "GSE115448['locus_tag'] = IDs\n",
    "GSE115448 = GSE115448.drop(['GeneSymbol','GeneName'], axis = 1)\n",
    "GSE115448 = GSE115448[['locus_tag']+sample_list]\n",
    "print(GSE115448.shape)\n",
    "GSE115448.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c4cd240",
   "metadata": {},
   "source": [
    "## L. plantarum response to acidic conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b9c586c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3174, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>con_1</th>\n",
       "      <th>con_2</th>\n",
       "      <th>pH5.0_1</th>\n",
       "      <th>pH5.0_2</th>\n",
       "      <th>pH5.5_1</th>\n",
       "      <th>pH5.5_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>423.10</td>\n",
       "      <td>406.42</td>\n",
       "      <td>619.47</td>\n",
       "      <td>640.82</td>\n",
       "      <td>468.07</td>\n",
       "      <td>465.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>672.64</td>\n",
       "      <td>645.10</td>\n",
       "      <td>897.24</td>\n",
       "      <td>916.21</td>\n",
       "      <td>681.56</td>\n",
       "      <td>685.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>510.91</td>\n",
       "      <td>559.08</td>\n",
       "      <td>499.21</td>\n",
       "      <td>448.14</td>\n",
       "      <td>522.82</td>\n",
       "      <td>514.83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>358.58</td>\n",
       "      <td>345.58</td>\n",
       "      <td>288.61</td>\n",
       "      <td>290.84</td>\n",
       "      <td>294.79</td>\n",
       "      <td>299.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>437.95</td>\n",
       "      <td>431.14</td>\n",
       "      <td>354.50</td>\n",
       "      <td>360.22</td>\n",
       "      <td>357.30</td>\n",
       "      <td>361.36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag   con_1   con_2  pH5.0_1  pH5.0_2  pH5.5_1  pH5.5_2\n",
       "0   lp_0001  423.10  406.42   619.47   640.82   468.07   465.00\n",
       "1   lp_0002  672.64  645.10   897.24   916.21   681.56   685.90\n",
       "2   lp_0004  510.91  559.08   499.21   448.14   522.82   514.83\n",
       "3   lp_0005  358.58  345.58   288.61   290.84   294.79   299.09\n",
       "4   lp_0006  437.95  431.14   354.50   360.22   357.30   361.36"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE143834 = pd.read_csv('../data/transcriptomics/GSE143834_Lactobacillus_plantarum_gene_expression_pH_.csv')\n",
    "GSE143834 = GSE143834[['Gene name','Start','End','Strand',\n",
    "                       'con_1','con_2','pH5.0_1','pH5.0_1.1','pH5.5_1','pH5.5_2']]\n",
    "GSE143834 = GSE143834.rename(columns = {'pH5.0_1.1':'pH5.0_2'})\n",
    "IDs = []\n",
    "for i in range(len(GSE143834.index)):\n",
    "    name = list(GSE143834['Gene name'])[i]\n",
    "    if 'lp_' in name:\n",
    "        IDs.append(name)\n",
    "    elif name in list(ref_features['symbol']):\n",
    "        temp = ref_features[ref_features['symbol'] == name ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    elif name in list(gb_features['symbol']):\n",
    "        temp = gb_features[gb_features['symbol'] == name ]\n",
    "        IDs.append( list(temp['locus_tag'])[0] )\n",
    "    else:\n",
    "        IDs.append(name)\n",
    "GSE143834['locus_tag'] = IDs\n",
    "GSE143834 = GSE143834[['locus_tag','con_1','con_2','pH5.0_1','pH5.0_2','pH5.5_1','pH5.5_2']]\n",
    "print(GSE143834.shape)\n",
    "GSE143834.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ec8fabd",
   "metadata": {},
   "source": [
    "## L. plantarum different carbon sources\n",
    "### Convert count to RPKM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "10225a0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3058, 23)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>LP_GLC_rep1</th>\n",
       "      <th>LP_GLC_rep2</th>\n",
       "      <th>LP_PAC1_rep1</th>\n",
       "      <th>LP_PAC1_rep2</th>\n",
       "      <th>LP_PAC2_rep1</th>\n",
       "      <th>LP_PAC2_rep2</th>\n",
       "      <th>LP_FOSPAC1_rep1</th>\n",
       "      <th>LP_FOSPAC1_rep2</th>\n",
       "      <th>LP_FOSPAC2_rep1</th>\n",
       "      <th>...</th>\n",
       "      <th>LP_HMOPAC1_rep1</th>\n",
       "      <th>LP_HMOPAC1_rep2</th>\n",
       "      <th>LP_HMOPAC2_rep1</th>\n",
       "      <th>LP_HMOPAC2_rep2</th>\n",
       "      <th>LP_XG_rep1</th>\n",
       "      <th>LP_XG_rep2</th>\n",
       "      <th>LP_XGPAC1_rep1</th>\n",
       "      <th>LP_XGPAC1_rep2</th>\n",
       "      <th>LP_XGPAC2_rep1</th>\n",
       "      <th>LP_XGPAC2_rep2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lp_0001</td>\n",
       "      <td>310.016351</td>\n",
       "      <td>378.385267</td>\n",
       "      <td>186.335784</td>\n",
       "      <td>181.704531</td>\n",
       "      <td>174.767584</td>\n",
       "      <td>331.888344</td>\n",
       "      <td>274.911717</td>\n",
       "      <td>307.043525</td>\n",
       "      <td>133.515007</td>\n",
       "      <td>...</td>\n",
       "      <td>502.441837</td>\n",
       "      <td>465.934221</td>\n",
       "      <td>223.878364</td>\n",
       "      <td>190.776541</td>\n",
       "      <td>145.272896</td>\n",
       "      <td>114.711457</td>\n",
       "      <td>106.658059</td>\n",
       "      <td>150.624559</td>\n",
       "      <td>168.328237</td>\n",
       "      <td>192.070155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lp_0002</td>\n",
       "      <td>669.481193</td>\n",
       "      <td>750.643207</td>\n",
       "      <td>618.965397</td>\n",
       "      <td>640.799551</td>\n",
       "      <td>468.851070</td>\n",
       "      <td>484.447600</td>\n",
       "      <td>459.368611</td>\n",
       "      <td>485.732790</td>\n",
       "      <td>602.754007</td>\n",
       "      <td>...</td>\n",
       "      <td>713.449932</td>\n",
       "      <td>623.604566</td>\n",
       "      <td>609.847872</td>\n",
       "      <td>572.761569</td>\n",
       "      <td>514.850531</td>\n",
       "      <td>449.567228</td>\n",
       "      <td>297.530491</td>\n",
       "      <td>352.205911</td>\n",
       "      <td>553.273363</td>\n",
       "      <td>581.633621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lp_0004</td>\n",
       "      <td>987.389048</td>\n",
       "      <td>994.441856</td>\n",
       "      <td>289.907038</td>\n",
       "      <td>324.869788</td>\n",
       "      <td>542.896017</td>\n",
       "      <td>391.251387</td>\n",
       "      <td>465.814018</td>\n",
       "      <td>456.112576</td>\n",
       "      <td>390.274635</td>\n",
       "      <td>...</td>\n",
       "      <td>307.528012</td>\n",
       "      <td>351.971933</td>\n",
       "      <td>528.534823</td>\n",
       "      <td>521.880441</td>\n",
       "      <td>269.890260</td>\n",
       "      <td>310.231795</td>\n",
       "      <td>480.214011</td>\n",
       "      <td>470.846952</td>\n",
       "      <td>397.137958</td>\n",
       "      <td>350.071757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lp_0005</td>\n",
       "      <td>114.401974</td>\n",
       "      <td>106.758145</td>\n",
       "      <td>34.261741</td>\n",
       "      <td>33.250165</td>\n",
       "      <td>62.854717</td>\n",
       "      <td>71.754663</td>\n",
       "      <td>190.020699</td>\n",
       "      <td>190.082872</td>\n",
       "      <td>74.980397</td>\n",
       "      <td>...</td>\n",
       "      <td>72.678780</td>\n",
       "      <td>80.916495</td>\n",
       "      <td>124.571592</td>\n",
       "      <td>74.847756</td>\n",
       "      <td>53.047054</td>\n",
       "      <td>90.030381</td>\n",
       "      <td>126.315493</td>\n",
       "      <td>114.415809</td>\n",
       "      <td>69.334865</td>\n",
       "      <td>85.637680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lp_0006</td>\n",
       "      <td>87.535495</td>\n",
       "      <td>76.947033</td>\n",
       "      <td>86.446226</td>\n",
       "      <td>82.303207</td>\n",
       "      <td>108.122124</td>\n",
       "      <td>100.719180</td>\n",
       "      <td>395.379766</td>\n",
       "      <td>374.195643</td>\n",
       "      <td>172.343487</td>\n",
       "      <td>...</td>\n",
       "      <td>61.886886</td>\n",
       "      <td>61.782763</td>\n",
       "      <td>102.612308</td>\n",
       "      <td>42.489197</td>\n",
       "      <td>53.565249</td>\n",
       "      <td>114.534706</td>\n",
       "      <td>183.798398</td>\n",
       "      <td>146.368833</td>\n",
       "      <td>60.189621</td>\n",
       "      <td>78.325210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  locus_tag  LP_GLC_rep1  LP_GLC_rep2  LP_PAC1_rep1  LP_PAC1_rep2  \\\n",
       "0   lp_0001   310.016351   378.385267    186.335784    181.704531   \n",
       "1   lp_0002   669.481193   750.643207    618.965397    640.799551   \n",
       "2   lp_0004   987.389048   994.441856    289.907038    324.869788   \n",
       "3   lp_0005   114.401974   106.758145     34.261741     33.250165   \n",
       "4   lp_0006    87.535495    76.947033     86.446226     82.303207   \n",
       "\n",
       "   LP_PAC2_rep1  LP_PAC2_rep2  LP_FOSPAC1_rep1  LP_FOSPAC1_rep2  \\\n",
       "0    174.767584    331.888344       274.911717       307.043525   \n",
       "1    468.851070    484.447600       459.368611       485.732790   \n",
       "2    542.896017    391.251387       465.814018       456.112576   \n",
       "3     62.854717     71.754663       190.020699       190.082872   \n",
       "4    108.122124    100.719180       395.379766       374.195643   \n",
       "\n",
       "   LP_FOSPAC2_rep1  ...  LP_HMOPAC1_rep1  LP_HMOPAC1_rep2  LP_HMOPAC2_rep1  \\\n",
       "0       133.515007  ...       502.441837       465.934221       223.878364   \n",
       "1       602.754007  ...       713.449932       623.604566       609.847872   \n",
       "2       390.274635  ...       307.528012       351.971933       528.534823   \n",
       "3        74.980397  ...        72.678780        80.916495       124.571592   \n",
       "4       172.343487  ...        61.886886        61.782763       102.612308   \n",
       "\n",
       "   LP_HMOPAC2_rep2  LP_XG_rep1  LP_XG_rep2  LP_XGPAC1_rep1  LP_XGPAC1_rep2  \\\n",
       "0       190.776541  145.272896  114.711457      106.658059      150.624559   \n",
       "1       572.761569  514.850531  449.567228      297.530491      352.205911   \n",
       "2       521.880441  269.890260  310.231795      480.214011      470.846952   \n",
       "3        74.847756   53.047054   90.030381      126.315493      114.415809   \n",
       "4        42.489197   53.565249  114.534706      183.798398      146.368833   \n",
       "\n",
       "   LP_XGPAC2_rep1  LP_XGPAC2_rep2  \n",
       "0      168.328237      192.070155  \n",
       "1      553.273363      581.633621  \n",
       "2      397.137958      350.071757  \n",
       "3       69.334865       85.637680  \n",
       "4       60.189621       78.325210  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE160565 = pd.read_csv('../data/transcriptomics/GSE160565_raw_gene_count_matrix.txt',sep='\\t')\n",
    "GSE160565 = GSE160565.dropna()\n",
    "GSE160565 = GSE160565.rename(columns={'Unnamed: 0':'gene'})\n",
    "sample_list = list(GSE160565.columns)[1:]\n",
    "sample_names = [\"LP_GLC_rep1\",\"LP_GLC_rep2\",\"LP_PAC1_rep1\",\"LP_PAC1_rep2\",\"LP_PAC2_rep1\",\"LP_PAC2_rep2\",\"LP_FOSPAC1_rep1\",\"LP_FOSPAC1_rep2\",\n",
    " \"LP_FOSPAC2_rep1\",\"LP_FOSPAC2_rep2\",\"LP_HMO_rep1\",\"LP_HMO_rep2\",\"LP_HMOPAC1_rep1\",\"LP_HMOPAC1_rep2\",\n",
    " \"LP_HMOPAC2_rep1\",\"LP_HMOPAC2_rep2\",\"LP_XG_rep1\",\"LP_XG_rep2\",\"LP_XGPAC1_rep1\",\"LP_XGPAC1_rep2\",\n",
    " \"LP_XGPAC2_rep1\",\"LP_XGPAC2_rep2\"]\n",
    "rename_dict = {sample_list[i]:sample_names[i] for i in range(len(sample_names))}\n",
    "GSE160565 = GSE160565.rename(columns=rename_dict)\n",
    "sample_list = list(GSE160565.columns)[1:]\n",
    "totalReads = {sample: sum(GSE160565[sample]) for sample in sample_list }\n",
    "IDs = []\n",
    "for i in range(len( GSE160565.index )):\n",
    "    gene_id = list(GSE160565['gene'])[i].split('-')[1].strip()\n",
    "    if 'RNA' in list(GSE160565['gene'])[i]:\n",
    "        IDs.append('-')\n",
    "    elif ( gene_id in list(ref_features['locus_tag']) ) or ( gene_id in list(gb_features['locus_tag']) ):\n",
    "        IDs.append( gene_id )\n",
    "    else:\n",
    "        IDs.append('-')\n",
    "GSE160565['locus_tag'] = IDs\n",
    "GSE160565 = GSE160565[GSE160565['locus_tag'] != '-']\n",
    "lengths=[]\n",
    "for i in range(len( GSE160565.index )):\n",
    "    gene_id = list(GSE160565['locus_tag'])[i]\n",
    "    if gene_id in list(ref_features['locus_tag']):\n",
    "        temp = ref_features[ref_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )\n",
    "    else:\n",
    "        temp = gb_features[gb_features['locus_tag'] == gene_id  ]\n",
    "        lengths.append( list(temp['feature_interval_length'])[0] )       \n",
    "GSE160565['length'] = lengths\n",
    "\n",
    "for sample in sample_list:\n",
    "    temp_rpkm = [ get_rpkm( list(GSE160565[sample])[i],\n",
    "                    list(GSE160565['length'])[i], totalReads[sample]) for i in range(len(GSE160565.index))]\n",
    "    GSE160565[sample] = temp_rpkm\n",
    "GSE160565 = GSE160565[['locus_tag']+sample_list]\n",
    "print(GSE160565.shape)\n",
    "GSE160565.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52cada60",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "ed85a99c",
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_merged = reduce(lambda  left,right: pd.merge(left,right,on=['locus_tag'],\n",
    "                            how='inner'), [GSE124050, GSE115448, GSE143834, GSE160565 ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "52e19a48",
   "metadata": {},
   "outputs": [],
   "source": [
    "geo_merged.to_csv('../data/transcriptomics/merged_rpkm.csv', index = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d08e9974",
   "metadata": {},
   "source": [
    "# Regulons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b9636d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_regprecise(file):\n",
    "    regulon = {}\n",
    "    ffile = open(file, \"rt\")\n",
    "    lines = ffile.readlines()\n",
    "    ffile.close()\n",
    "    for line in lines:\n",
    "        line = line.strip()\n",
    "        if len(line)<1:\n",
    "            continue\n",
    "        elif '#' in line:\n",
    "            RF = ((line.split('-')[1]).split(':')[0]).strip()\n",
    "            regulon[ RF ] = []\n",
    "        else:\n",
    "            g = str(line.split('\\t')[2]).strip()\n",
    "            regulon[RF].append(g)\n",
    "    return regulon "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "157046c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# regulons={}\n",
    "# regulons['Lbrevis'] = read_regprecise('../data/regulon/Regulons_Lbrevis_ATCC_367.txt')\n",
    "# regulons['Lplantarum'] = read_regprecise('../data/regulon/Regulons_Lplantarum_WCFS1.txt')\n",
    "# regulons['Lrhamnosus'] = read_regprecise('../data/regulon/Regulons_Lrhamnosus_GG.txt')\n",
    "# dump_pickle(regulons, '../data/regulon/regulons.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7966a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "regulons = load_pickle('../data/regulon/regulons.pkl')\n",
    "print(regulons['Lplantarum'].keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d3840a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
